File: solutions/3915-maximum-product-of-two-integers-with-no-common-bits/maximum-product-of-two-integers-with-no-common-bits.cpp
class Solution {
public:
    long long maxProduct(vector<int>& nums) {
        long long maxnum=*max_element(nums.begin(), nums.end()); //get largest element
        long long bits=0; //num of bits in that element
        while((1<<bits)<=maxnum){
            bits++;
        }
        int num_of_mask=1<<bits; //how many possible masks are there (if bits=3, then 8 masks)

        vector<int>best_mask(num_of_mask,0);//largest number corresponding to each mask
        for(int v:nums){
            best_mask[v]=max(best_mask[v],v);
        }

        //if best_mask[010]=5, then best_mask[110]=5 as 010 is subset of 110 (010 & 110= 010)
        for(int b=0;b<bits;b++){
            for(int mask=0 ;mask<num_of_mask; mask++){
                if(mask & (1<<b)){ //if bit b of a mask is set
                    best_mask[mask]=max(best_mask[mask], best_mask[mask ^ (1<<b)]); //taking xor will make that bit unset so that we get a subset of mask
                }
            }
        }
        long long ans=0;
        long long allmask=num_of_mask-1; //111
        for(long long v:nums){
            long long complement=allmask^v; //gives all mask that donot overlap with v

            if(best_mask[complement]>0 and best_mask[complement]!=v){
                ans=max(ans, v* best_mask[complement]);
            }
        }
        return ans;
    }
};

File: solutions/3915-maximum-product-of-two-integers-with-no-common-bits/maximum-product-of-two-integers-with-no-common-bits.java
class Solution {
    public long maxProduct(int[] nums) {
        int max=0;
        for(int val:nums)max=Math.max(max,val);
        int totalBit=(int)(Math.log(max)/Math.log(2) + 1);
        int maxMask=(1<<totalBit)-1;

        /*dp[val] tells if we represent val in binary number then all set bit everytime goes to unset
        then from all subset take max of them and store in dp[val]
        eg: comp=0101 take all combination of set bit to unset means place 1->0 or no change(means itself)
        [0100,0001,0000,itself 0101]*/
        int[] dp=new int[maxMask+1];
        for(int val:nums)dp[val]=val;
        
        for(int i=0; i<totalBit; i++){
            for(int val=1; val<=maxMask; val++){
                /*every time every set bit goes to unset and we take maximum from all of them*/
                if((val & (1<<i) )!=0){//only for set bit we need to unset
                    int offBitNumber=(val ^ (1<<i));//set bit -> unset
                    dp[val]=Math.max(dp[val],dp[offBitNumber]);
                }
            }
        }
        long ans=0;
        for(int val:nums){
            /*take complement of val and look for dp[comp] which tells all subset including itself 
            gives max where in every subset every alter set bit goes to unset
            eg: val=1010 comp=0101 take all combination of set bit in comp means place 1->0
            [0100,0001,0000,itself 0101]*/ 
            int comp=(maxMask ^ val);
            ans=Math.max(ans,(long)val*dp[comp]);
        }
        return ans;
    }
} 
//Time=> O(N∗Log(Max(Nums)))
//Space=> O(Max(Nums))